### PR TITLE
Misc cleanups around `Netty4HttpPipeliningHandler`

### PR DESCRIPTION
**Pull request by DaveCTurner**
- Make `Netty4HttpResponse` sealed, we need to know all its impls
- Rename `Netty4FullHttpResponse` to contrast with chunked response
- Rename `doWrite` overloads
- Reorder `finishChunkedWrite()`, we're ready to handle the next
  response before completing the promise
- Add missing test for splitting large responses
- Add a few extra assertions
- Clean up IDE warnings

### CODE REVIEW

File: modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandlerTests.java [hunk 1]
```diff
@@ -211,6 +215,43 @@ public void testPipeliningRequestsAreReleased() throws InterruptedException {
...
+        embeddedChannel.writeInbound(createHttpRequest("/test"));
+        final Netty4HttpRequest request = embeddedChannel.readInbound();
+        final var maxSize = (int) NettyAllocator.suggestedMaxAllocationSize() / 2;
+        final var content = new ZeroBytesReference(between(0, maxSize));
+        final var response = request.createResponse(RestStatus.OK, content);
+        final var promise = embeddedChannel.newPromise();
+        embeddedChannel.writeAndFlush(response, promise);
+        assertTrue(promise.isDone());
+        assertThat(messagesSeen, hasSize(1));
+        assertSame(response, messagesSeen.get(0));
```
Comment (pxsalehi) [inline modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandlerTests.java:230]: should we also assert that the seen message is an instance of full HTTP response?

**Review by pxsalehi — APPROVED** [review]
LGTM (although I am still getting to learn this area, so it might worth waiting for a second one).

File: modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandlerTests.java (hunk 1 above)
Comment (DaveCTurner) [reply modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandlerTests.java:232]: Yep can do, see 5d30ae9df72

**Review by ywangd — APPROVED** [review]
LGTM

**Review by Tim-Brooks — APPROVED** [review]
LGTM

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4RestResponse.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 import org.elasticsearch.http.HttpResponse;
 
-public interface Netty4RestResponse extends HttpResponse, HttpMessage {
+public sealed interface Netty4RestResponse extends HttpResponse, HttpMessage permits Netty4FullHttpResponse, Netty4ChunkedHttpResponse {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4RestResponse.java:15]: Yeah it was a little noisy to do that in this same PR, I opened https://github.com/elastic/elasticsearch/pull/104675.

**Comment by DaveCTurner** [comment]
FWIW this was opened to prepare things for https://github.com/elastic/elasticsearch/pull/104851.

**Pull request merged**